Adding UQ for TIE and electrical losses

nbodini opened the pull request:
Added 
electrical_losses_uq.py 
turbine_long_term_gross_energy_uq.py
which add UQ to the calculation of electrical losses and long-term gross energy calculations, based on a Monte Carlo approach.

ejsimley commented on operational_analysis/methods/electrical_losses_uq.py:
```diff
@@ -0,0 +1,252 @@
+
+        Args:
+         plant(:obj:`PlantData object`): PlantData object from which EYAGapAnalysis should draw data.
+
```
the new UQ-related args should be described here in the docstring

nbodini replied:
I will add the following description:

        """
        Initialize electrical losses class with input parameters

        Args:
         plant(:obj:`PlantData object`): PlantData object from which EYAGapAnalysis should draw data.
         num_sim:(:obj:`int`): number of Monte Carlo simulations
         uncertainty_meter(:obj:`float`): uncertainty imposed to revenue meter data
         uncertainty_scada(:obj:`float`): uncertainty imposed to scada data
         uncertainty_correction_threshold(:obj:`float`): The data availability threshold (fraction) 
                                                         under which months should be eliminated (default is 0.95)
        """

ejsimley commented on operational_analysis/methods/turbine_long_term_gross_energy_uq.py:
```diff
@@ -0,0 +1,535 @@
+        Args:
+         plant(:obj:`PlantData object`): PlantData object from which TurbineLongTermGrossEnergy should draw data.
+         num_sim:(:obj:`int`): number of Monte Carlo simulations
+         uncertainty_meter(:obj:`float`): uncertainty imposed to revenue meter data
```
uncertainty_meter isn't used in this class, right? So it should be removed from the arguments.

nbodini replied:
Right thanks, I will remove that!

jordanperr replied:
Has this been resolved?

nbodini replied:
Yes, I've now marked this as resolved.

jordanperr commented:
Before a more detailed review, I see two issues right off the bat:

- There are no unit tests with this pull request (required).
- There are no example notebooks exercising the methods (optional).

Please let me know if I can assist you with writing these tests. Here is an example test which you can use as a reference: https://github.com/NREL/OpenOA/blob/master/test/int_pruf_plant_analysis.py

The instructions for running tests can be found in the readme.

nbodini commented:
I have added the first set of modifications to the AEP calculation. This now includes the ability for the user to choose which temporal frequency to use (daily, monthly) and which variables to use as regression inputs (wind speed, wind direction, temperature).
Examples and tests for this modification are currently missing, and waiting to have a good dataset to use for this purpose.

(I have tried to create a separate PR for this additional modification, but I wasn't able to - maybe because I already had the UQ one still pending review?)

nbodini commented:
Following the discussion in PR #64, I have now removed the original versions of the scripts and tests, and left the updated ones only (and updated their names).

ejsimley commented on operational_analysis/methods/electrical_losses.py:
```diff
@@ -28,16 +36,35 @@ class ElectricalLosses(object):
          plant(:obj:`PlantData object`): PlantData object from which EYAGapAnalysis should draw data.
-
+         num_sim:(:obj:`int`): number of Monte Carlo simulations
+         UQ:(:obj:`str`): choice whether to perform ('Y') or not ('N') uncertainty quantification
```
Can you change this to a boolean - e.g., True is UQ?

nbodini replied:
Changed!

ejsimley commented on operational_analysis/methods/electrical_losses.py:
```diff
@@ -28,16 +36,35 @@ class ElectricalLosses(object):
+         UQ:(:obj:`str`): choice whether to perform ('Y') or not ('N') uncertainty quantification
+         uncertainty_meter(:obj:`float`): uncertainty imposed to revenue meter data
+         uncertainty_scada(:obj:`float`): uncertainty imposed to scada data
+         uncertainty_correction_threshold(:obj:`float`): The interval of data availability thresholds (fractions) 
```
This is a tuple of floats, instead of a single float. Can you update the data type in the docstring to reflect this?

nbodini replied:
Changed!

ejsimley commented on operational_analysis/methods/turbine_long_term_gross_energy.py:
```diff
@@ -42,15 +47,36 @@ class TurbineLongTermGrossEnergy(object):
         Args:
          plant(:obj:`PlantData object`): PlantData object from which TurbineLongTermGrossEnergy should draw data.
-         max_power_filter(float): 
+         UQ:(:obj:`str`): choice whether to perform ('Y') or not ('N') uncertainty quantification
```
Similarly, changing this to a boolean seems simpler for specifying UQ.

nbodini replied:
Changed!

ejsimley commented on operational_analysis/methods/turbine_long_term_gross_energy.py:
```diff
@@ -42,15 +47,36 @@ class TurbineLongTermGrossEnergy(object):
+                                            should be applied (default 0.85)
+         uncertainty_wind_bin_thresh(:obj:`float`): The filter threshold for each bin (default is 2 m/s)
+         uncertainty_correction_threshold(:obj:`float`): The threshold (fraction) above which daily scada energy data
+                                                should be corrected (default is 0.90)
```
The last three arguments are tuples, but the docstring datatypes are only listed as floats.

nbodini replied:
...and changed!

nbodini commented on operational_analysis/methods/electrical_losses.py:
```diff
@@ -28,16 +36,35 @@ class ElectricalLosses(object):
 
     @logged_method_call
-    def __init__(self, plant, correction_thresh=.95):
+    def __init__(self, plant, UQ = True, num_sim = 20000, uncertainty_meter=0.005, uncertainty_scada=0.005,
```
I have switched to UQ = False by default.

nbodini commented on operational_analysis/methods/electrical_losses.py:
```diff
@@ -49,6 +76,11 @@ def __init__(self, plant, correction_thresh=.95):
         self._hours_per_day= 24 # Hours per day converter
+    
+        # Define relevant uncertainties, to be applied in Monte Carlo sampling
+        self.uncertainty_meter = np.float64(uncertainty_meter)
```
I have moved these definitions within the UQ == True clause. And I have done the same for the TIE class.

nbodini commented on operational_analysis/methods/electrical_losses.py:
```diff
@@ -49,6 +76,11 @@ def __init__(self, plant, correction_thresh=.95):
+    
+        # Define relevant uncertainties, to be applied in Monte Carlo sampling
+        self.uncertainty_meter = np.float64(uncertainty_meter)
+        self.uncertainty_scada = np.float64(uncertainty_scada)
```
I have removed that.

nbodini commented on operational_analysis/methods/turbine_long_term_gross_energy.py:
```diff
@@ -42,15 +47,36 @@ class TurbineLongTermGrossEnergy(object):
 
     @logged_method_call
-    def __init__(self, plant):
+    def __init__(self, plant, UQ = True, num_sim = 100, uncertainty_scada=0.005,
```
I have switched to UQ = False by default, and added a comment on why the default value for num_sim is lower than what chosen in the electrical_losses.

nbodini commented on operational_analysis/methods/turbine_long_term_gross_energy.py:
```diff
@@ -66,7 +92,7 @@ def __init__(self, plant):
-              
+
         # Dictionary to convert time interval indicator into minutes
         self._time_conversion = {'10T': 10.,
```
Changed.

nbodini commented on operational_analysis/methods/electrical_losses.py:
```diff
@@ -28,16 +36,35 @@ class ElectricalLosses(object):
+        
         self._plant = plant
         
         self._time_conversion = {'1T': 1.,
```
Changed.

nbodini commented on operational_analysis/methods/turbine_long_term_gross_energy.py:
```diff
@@ -75,27 +101,25 @@ def __init__(self, plant):
         
-
+        # Define relevant uncertainties, to be applied in Monte Carlo sampling
+        self.uncertainty_scada = np.float64(uncertainty_scada)
```
Removed.

nbodini commented on operational_analysis/methods/turbine_long_term_gross_energy.py:
```diff
@@ -75,27 +101,25 @@ def __init__(self, plant):
     @logged_method_call
-    def run(self, reanal_subset = ['erai', 'ncep2', 'merra2'], max_power_filter = 0.85, 
-            wind_bin_thresh = 2, correction_threshold = 0.90, enable_plotting = False,
+    def run(self, reanal_subset = ['erai', 'ncep2', 'merra2'], enable_plotting = False,
```
I have moved this to init.

nbodini commented on operational_analysis/methods/turbine_long_term_gross_energy.py:
```diff
@@ -75,27 +101,25 @@ def __init__(self, plant):
 
     @logged_method_call
-    def run(self, reanal_subset = ['erai', 'ncep2', 'merra2'], max_power_filter = 0.85, 
-            wind_bin_thresh = 2, correction_threshold = 0.90, enable_plotting = False,
```
I have checked with @moptis that there are currently no examples that use this class, so this change should be safe.
